src/main/java/mezz/jei/transfer/RecipeTransferUtil.java

```diff
@@ -67,66 +71,158 @@ public static boolean transferRecipe(RecipeTransferManager recipeTransferManager
+		}
 	}
 
 	public static boolean allowsTransfer(@Nullable IRecipeTransferError error) {
 		return error == null ||
 			error.getType() == IRecipeTransferError.Type.COSMETIC;
 	}
 
-	public static class MatchingItemsResult {
-		public final Map<Integer, Integer> matchingItems = new HashMap<>();
-		public final List<Integer> missingItems = new ArrayList<>();
+	public static boolean validateSlots(
+		Player player,
+		Collection<TransferOperation> transferOperations,
+		Collection<Slot> craftingSlots,
+		Collection<Slot> inventorySlots
+	) {
+		Set<Integer> inventorySlotIndexes = inventorySlots.stream()
+			.map(s -> s.index)
+			.collect(Collectors.toSet());
+		Set<Integer> craftingSlotIndexes = craftingSlots.stream()
+			.map(s -> s.index)
+			.collect(Collectors.toSet());
+
+		// check that all craftingTargetSlots are included in craftingSlots
+		{
+			List<Integer> invalidRecipeIndexes = transferOperations.stream()
+				.map(TransferOperation::craftingSlot)
+				.map(s -> s.index)
+				.filter(s -> !craftingSlotIndexes.contains(s))
+				.toList();
+			if (!invalidRecipeIndexes.isEmpty()) {
+				LOGGER.error(
+					"Transfer handler has invalid slots for the destination of the recipe, " +
+						"the slots are not included in the list of crafting slots. " +
+						StringUtil.intsToString(invalidRecipeIndexes)
+				);
+				return false;
+			}
+		}
+
+		// check that all ingredientTargetSlots are included in inventorySlots or recipeSlots
+		{
+			List<Integer> invalidInventorySlotIndexes = transferOperations.stream()
+				.map(TransferOperation::inventorySlot)
+				.map(s -> s.index)
+				.filter(s -> !inventorySlotIndexes.contains(s) && !craftingSlotIndexes.contains(s))
+				.toList();
+			if (!invalidInventorySlotIndexes.isEmpty()) {
+				LOGGER.error(
+					"Transfer handler has invalid source slots for the inventory stacks for the recipe, " +
+						"the slots are not included in the list of inventory slots or recipe slots. " +
+						StringUtil.intsToString(invalidInventorySlotIndexes) + 
+						"\n inventory slots: " + StringUtil.intsToString(inventorySlotIndexes) +
+						"\n crafting slots: " + StringUtil.intsToString(craftingSlotIndexes)
+				);
+				return false;
+			}
+		}
+
+		// check that crafting slots and inventory slots do not overlap
+		{
+			Set<Integer> overlappingSlots = inventorySlotIndexes.stream()
+				.filter(craftingSlotIndexes::contains)
+				.collect(Collectors.toSet());
+			if (!overlappingSlots.isEmpty()) {
+				LOGGER.error(
+					"Transfer handler has invalid slots, " +
+						"inventorySlots and craftingSlots should not share any slot, but both have: " +
+						StringUtil.intsToString(overlappingSlots)
+				);
+				return false;
+			}
+		}
+
+		// check that all slots can be picked up by the player
+		{
+			List<Integer> invalidPickupSlots = Stream.concat(
+					craftingSlots.stream(),
+					inventorySlots.stream()
+				)
+				.filter(slot -> !slot.mayPickup(player))
```

Comment:
This check is technically a breaking change as it causes one of the containers I have an `IRecipeTransferInfo` to no longer work as if the slot is empty I return false for `mayPickup`. Previously (and in the remaining places in `BasicRecipeTransferHandler`, though I am slightly unsure about the Server calls) before checking and failing if unable to pickup from a slot it was checked that the slot is not empty.

Reply:
That makes sense. I will change this so that it only checked `mayPickup` on slots that contain something.